### ISSUE TITLE
Improving Startup and Overall Performance

### ISSUE THREAD

**Issue by rkoeninger** [rkoeninger/ShenScript#7]
# Check List

  - [x] Hoist idle symbols, global function and symbol references
  - [x] Optimize specific commonly used functions in both sync and async form
    - [x] `shen.mod`, `integer?`, `@p`, `shen.numbyte`, `shen.byte->digit`
    - [x] `symbol?`, `variable?`, `shen.pvar?`
    - [x] `shen.fillvector`
    - [x] `shen.dict` functions
  - [x] Bring in the dynamic eval expansion (when kernel 22.0 is released)
  - [x] Find a way to cut down on the amount of `async/await` used in the async kernel (Firefox load time is 10x faster without async)
  - [ ] Used fixed-arity application where possible, avoid rest arguments in wrapper functions created by `fun`

# Original Post

Picking up on the performance issue mentioned in #5

**Observations**

Startup performance in async mode is unacceptably slow. Seems to vary mainly with the browser. Brave creates the kernel in about 1.2s, but Firefox takes about 15s.

@vrescobar reported a 25s startup time in Chrome.

I think startup needs to happen in at most 500ms for ShenScript to be feasible for production use.

Startup time is normally much lower in sync mode, meaning async/await performance likely differs considerably between JavaScript engines.

**Solutions**

Some work has already been done to reduce the number of functions that are actually async and then don't get awaited to cut down on unnecessary asynchronicity. More opportunities for this could be done by analysis at build time or a manual survey and just listing the function names by configuration.

A more drastic approach would be to abandon async/await, even though it's convenient, I've heard it's slower than writing callback-style code. That would require a much more involved transpiler.

**Comment by vrescobar** [comment]
I didn't read the codebase, but I can tell that being javascript single threaded, Async will be the right answer for:
- IO
- Handling browser events
- Slow computations

In the case of Shen Script I believe that this job should be delegated to the guest language which will also offer constructs to handle IO/Events, and if the code does not reflect that it will end up freezing the UI, but everything should be up to shen's code I believe.

**Comment by rkoeninger** [comment]
@vrescobar One of the sections of the docs I'm in the process of writing is the explanation of why async/await was needed.

The primitive IO functions in Shen are synchronous. The `read-byte` function in particular makes porting Shen to JavaScript difficult. Code that uses `read-byte`, including kernel functions like `read-file-as-charlist` and `load` are written in synchronous style. So we need a `read-byte` function that is asynchronous in the host language but still looks like a synchronous function in the guest language.

I was thinking about this and realized "that's what async/await does!" So `read-byte` and other IO functions are `async` functions or return `Promise`s. And so the functions that call the primitive IO functions need to be `async` so they can `await` the primitive IO functions. And then functions that call *those* functions need to be `async`, and so on, all the way to the repl function, `shen.shen`.

So I don't think you can deal with asynchronicity in the guest language because the guest language was designed in a way that counts on the host language to deal with it.

**Comment by rkoeninger** [comment]
First draft of this async explanation now in the docs: https://shenscript.readthedocs.io/en/latest/design.html#optional-pervasive-async

**Comment by rkoeninger** [comment]
I also wrote about the hybrid sync/async approach here: https://shenscript.readthedocs.io/en/latest/design.html#hybrid-sync-async-kernel

**Comment by rkoeninger** [comment]
@vrescobar Also, I forgot to mention that having the chrome dev tools closed while loading the page increases speed by about 10x. It takes 15secs for me with the console open, 1.2secs with it closed.

**Comment by vrescobar** [comment]
you are right about the dev tools, I cannot understand why, maybe the chrome debugger has a high impact on the loadtime.

**Comment by vrescobar** [comment]
I've been trying the synchronous mode in the browser and it is consistently starting less than two seconds (1261ms) in the same setup from the other day (Chrome with or without devtools, huge laptop with many res).

That's very closed to the original goal.

On safari (iphone 6s) it takes 5942ms which is still a bit far from optimal. If I set webpack from development to production it improves to 3028ms which is a great improvement but still not reaching its goal (only on mobile by the way).

There must be a way to lighten the AST.

Checking the chrome performance tab, and looking at the profiler is very revealing. Basically `Cons`, `trampoline` and `throw` and the `...args` from funSync are taking the largest times. Maybe some small rewrites there could make the difference:

<img width="1437" alt="Screenshot 2019-08-22 at 00 04 27" src="https://user-images.githubusercontent.com/6759126/63471890-c874f480-c470-11e9-93cb-dced0a34f526.png">

**Comment by vrescobar** [comment]
I added a commit which improves significantly the performance on mobile, but it doesn't change much on browser. It contributes to the fix of this issue but doesn't completely solve it (we are not subsecond yet).

Regarding stop using `Object.freeze` I wouldn't worry too much since it is mostly a "security" contract and all the calls and access to this code happen within the shen engine and your code, so there is no risk that an uninformed user changes these objects while the trampoline is running.

Unfortunately github stacked my pull requests together since they depended on each other.

**Comment by vrescobar** [comment]
I tried some optimizations there too,
also without much success. My guess went to the argument serialization, using `...args` vs using `f.apply(args)` could trigger an additional transformation under the cover but it didn’t make any positive difference.

I still think that the Cons list and arg serialization play a big role and I would like to pass them by reference rather copy and convert the Cons to a more imperative structure but I didn’t get so far.

The trampoline code also had a good performance penalty but unless we rewrite the AST using gotos or loops it won’t get any better; I tried a Y-Combinator in life she but the performance seems largely impacted.

The other thing still to try would be to move some parts to wasm.

Víctor R. Escobar

**Comment by vrescobar** [comment]
These both are expensive ideas to implement, lots of refactoring, and I guess you are comfy but I am not that into the codebase since I have to work and I am learning official shen.

How do you measure performance? I think the "final" solution should be Web Assembly and custom data structures, but I cannot help right now with it since I don't know really know wasm (it is a lisp btw). Maybe we can start implementing a new persistent Cons based on js arrays and hardcode some common use cases, after all performance is an illusion.

**Comment by vrescobar** [comment]
What do you think about using clojurescript's immutable list? 

https://github.com/immutable-js/immutable-js

if we just import `{ List }` Webpack should do some dead code elimination, isn't it?

**Comment by tizoc** [comment]
@rkoeninger you will not see much difference in the time it takes to run the test suite because none of the tests are numerically heavy.

Btw, for some reason some of the code blocks are missing in the documentation, see the bottom of https://shenscript.readthedocs.io/en/latest/design.html for example (but Github does render them)

**Comment by rkoeninger** [comment]
The casts also got removed from all the non-numeric operations too (except converting between shen and js booleans), and right now we're just focusing on improving web startup time so a page using ShenScript won't take 15s to load.

PS - takes for the tip on the missing code blocks. Sphinx likes those to be labelled I guess.

**Comment by tizoc** [comment]
Sorry, didn't intend to derail things.

How do I compile a chunk of Kl code into Javascript? I have tried this, but I always get a "not a valid form error":

```
(js.ast.render (js.ast.compile [defun a [X] [+ X X]]))
(js.ast.render (js.ast.compile (hd (read-from-string "(defun a (X) (+ X X))"))))
```

I also see that the generated bundle has a lots of calls to `eval`, I guess this is webpack's doing.

**Comment by tizoc** [comment]
@rkoeninger it seems a good amount of time is being spent in the dictionary functions, something you can try is to override those with a native implementation of dicts using either objects or `Map` (the introduction of those as overrideable was made thinking about Javascript in particular because unlike in most other languages, no native hashing function is exposed by the platform), that is going to avoid a lot of try/catch blocks being executed and `raise`s. This in addition will speed up all uses of `get` and `put`, which should help with startup time.

**Comment by rkoeninger** [comment]
@tizoc Fixed `js.ast.compile` in latest master. It was expecting a tree of JS arrays as opposed to a tree of Shen/KL lists and needed a conversion function thrown in there.

**Comment by tizoc** [comment]
@rkoeninger thank you! it works now.

Another thing that I see taking up a lot of time is symbols. Unless the javascript compiler somehow optimizes this magically, I think compiling symbols to ```s`symbol-name` ``` is adding considerable overhead because each time code execution reaches a symbol it has to do a lookup to find the symbol, if you can figure a way to have the symbol be instantiated only once and then reference it by just referencing a variable that would help too. This is easy in Scheme and Common Lisp because both languages provide syntax for symbols, not sure what is the best way to handle it in Javascript.

**Comment by rkoeninger** [comment]
Symbols in JS are a little different from symbols in Lisp. JS symbols are all unique even if they have the same description string and they don't even necessarily have a description string. So `Symbol('abc') === Symbol('abc')` is false.

`Symbol.for(name)` will intern a symbol by a particular name and always return the same symbol for the same name. The ```s`symbol-name` ``` syntax turns into `Symbol.for('symbol-name')`. That's neat for user code, but could use `symbolOf` instead.

Or are you suggesting not to use JS symbols? I could add a `Sym` class with custom equality logic that compares names. They might be faster to create, but they would take longer to compare. I think JS symbols use an identity comparison that is about as fast as comparing two pointers.

**Comment by tizoc** [comment]
No, what I mean is that every time the execution path reaches a call to `Symbol.for`, it has to do a lookup, right? and then depending on if the symbol exists already, intern the string or return the already existing instance. This would be like compiling to Lisp/Shen that uses calls to `(intern "thesymbol")` instead of `'symbol`. Having to do a lookup from an input string is far more expensive than just referencing an address.

What I have in mind (not familiar enough with the compiler at this moment to know how viable it is btw) is to "hoist" symbols to top-level variables so that they can be referenced directly to avoid calls to `Symbol.for`:

```
$.f.define($.s`test`, $.s`X`, $.s`->`, $.r([$.s`symbol`, $.s`symbol`])))
```

would instead be:

```
const $SYMS$symbol = Symbol.for("symbol"); // Symbol.for is called only ever once
// then this variable is referenced every time we want that symbol
// only doing it for a single symbol here, but same would apply to test, X, etc
$.f.define($.s`test`, $.s`X`, $.s`->`, $.r([ $SYMS$symbol,  $SYMS$symbol])));
```

Maybe this is a dumb idea, not sure, I have to leave for a bit now but I'm going to write a test in plain javascript to see is `Symbol.for` does add overhead as I think or not and I will let you know.

**Comment by rkoeninger** [comment]
#### Interned Symbol Hoisting

Oh, I see. That would be most applicable in the kernel where there's a large block of pre-rendered code with many repeated symbols. If you did this for every expression, it wouldn't prevent many lookups.

I just tried a quick benchmark of looking up randomly-named symbols with `Symbol.for` and it appears to do about 17,000,000/sec in Chromium.

It also just occurred to me that it might be spending so much time in `s` because `s` calls `String.raw` which does toString conversions and interleaving and so a decent amount of conditional logic. When the symbol doesn't need to be a template string, might as well use `symbolOf` at least. That's an easy thing to try out.

**Comment by rkoeninger** [comment]
#### Pinhole Optimizations

I also haven't done anything with pinhole optimizations yet.

The `map` function looks ripe for optimization as I could have it not just use a JS loop instead of recursing through trampolines (and maybe a promise-trampoline chain), I could also take advantage of the fact that conses are mutable under the hood and build the list from the tail without having to reverse it or build a JS array of results and then iterate over it again to build a result list.

@tizoc In your experience, what are some of the best functions to pinhole-optimize?

**Comment by tizoc** [comment]
```javascript
let res = 0;
let arr = [0, 0, 0]
let sym = Symbol.for("string");

for (let i = 0; i < 99999999; i++) {
  res += 1;
  arr[i % 3] = Symbol.for("string");
  //arr[i % 3] = sym;
}

console.log(res);
console.log(arr);
```

On my laptop, the version that uses `Symbol.for` on every iteration of the loop runs in 2.58s, the one that references the variable to avoid symbol lookup runs in 0.37s. I am not sure how much of a problem this is in the current implementation, I just see in the profiler results that a lot of time is spent in `s`.

For optimizations, the `dict.*` functions (or just `hash`, but not doable in Javascript), and `get`  are good candidates (these last two depending on how expensive exception handling is). Then there are other functions that help with shen->kl compilation and typechecking speed, like `shen.pvar?`, `numbyte?`, `symbol?`, `shen.analyze-symbol`, etc, but I don't think those are affecting your load times.

See https://github.com/tizoc/shen-scheme/blob/master/src/overrides.shen

If exception handling adds overhead there are some things you can do when compiling the kernel code but that you can't enable for code in general because it is not safe: https://github.com/tizoc/shen-scheme/blob/master/src/compiler.shen#L128-L147

The transformation there makes use of `*/or` functions that return a default value instead of raising an exception, by doing so a bunch of functions get rewritten into a version that doesn't setup an exception handler.

Since the profiler results also shows that a lot of time is being spent on `raise` this is something that may help a bit too.

Note that I don't override `get` in `overrides.shen` because the above transformation takes care of avoiding the setup of an exception handler in `get`.

**Comment by tizoc** [comment]
@rkoeninger this version is as fast as the variable reference (not sure if it translates to the case where the table is much bigger):

```javascript
let res = 0;
let arr = [0, 0, 0];
let symbols = {};
symbols.string = Symbol.for("string");

for (let i = 0; i < 99999999; i++) {
  res += 1;
  arr[i % 3] = symbols.string;
}

console.log(res);
console.log(arr);

```

**Comment by rkoeninger** [comment]
I was resistant to the idea of binding symbol values to locals at first because it wasn't "neat" enough, but then I realized it could be done with the established iife pattern.

So if a function is going to refer to a symbol repeatedly, it can be wrapped in another function that takes the symbol value, refers to the parameter instead and then that outer function is immediately applied to arguments which are the looked-up values the function will need.

```shen
(defun sign (X)
  (cond ((> X 0) positive)
        ((< X 0) negative)
        (true    zero)))
```

becomes

```javascript
f.sign = ((positive$, negative$, zero$) =>
  X => X > 0 ? positive$ :
       X < 0 ? negative$ :
               zero$)
    (s`positive`, s`negative`, s`zero`);
```

The `X => ...` function is what gets assigned to `f.sign` and it has `positive$`, etc in scope with the interned symbol values so it shouldn't have to look them up again since they never change.

This can be taken a step further by inlining reference cells that hold onto global functions and symbols. This way, the function can still be redefined and existing referrers will still have a direct handle to where it's being stored. I originally thought the `$.functions.map` type of structure could be optimized by the JS runtime, but maybe it isn't I'll have to experiment with this. Each invocation of the function would still have to fetch the function from the reference cell but accessing index `0` on an array is probably faster than looking up a key in an object with 1000 properties.

I like this iife approach because it retains expression oriented syntax and allows for the possibility of memoizing the results of whatever pre-computation desired: symbol interns, global function reference cells, transpile-time constant expressions e.g. like building a long constant list once instead of every time:

https://github.com/Shen-Language/shen-sources/blob/304d7c15cd0a71ff11c161cdfb807a50d737ebc1/sources/sys.shen#L158-L164

A technique like this is used in ShenSharp which uses an AST interpreter where the enhanced AST holds onto the reference cell for global functions and symbols. I didn't think it would be relevant here since ShenScript isn't an interpreter, but here's an opportunity to do it here too.

Inlining pointer to reference cell that holds global function value:

https://github.com/rkoeninger/ShenSharp/blob/b57cfb9797d038c06de41b1d54009f0fee561979/Kl/Evaluator.fs#L72-L79

**Comment by tizoc** [comment]
I like that approach, seems like it would work quite well while at the same time being considerably easier to implement than what I had in mind of using variable references (which would require a quite different compilation model from what you have now).

**Comment by tizoc** [comment]
@rkoeninger I'm thinking that you can use a custom object type instead of an array and the JS compiler is probably going to be able to do a better job at optimizing it btw (like for example being able to avoid for out-of-bound accesses), will try it now and post results in a few minutes.

**Comment by tizoc** [comment]
Well, guess what, it is mostly the same in all cases, for reads both array and custom object reads are almost as fast as referencing a variable. It is only when there are writes that arrays get slower, I would have expected the opposite (reads more penalized than writes).

These are the tests that I did first, but they have too many writes, once I noticed that I was testing the wrong thing and removed the writes the times were 0.67s for variables and 0.71 for object and array.

Test with writes (but again, not what is needed in this case):

Plain version, finishes in 1.04s:

```javascript
let oldboxed = 0;
let boxed = 0;

for (let i = 0; i <= 999999999; i++) {
  oldboxed = boxed;
  boxed = i;
}

console.log(oldboxed);
console.log(boxed);
```

Array version, finishes in 1.95s:

```javascript
const oldboxed = [0];
const boxed = [0];

for (let i = 0; i <= 999999999; i++) {
  oldboxed[0] = boxed[0];
  boxed[0] = i;
}

console.log(oldboxed[0]);
console.log(boxed[0]);
```

Object version, finishes in 1.15s:

```javascript
function Box(value) {
  this.value = value;
}

Box.prototype.read = function () { return this.value; };
Box.prototype.write = function (value) { this.value = value };

const oldboxed = new Box(0);
const boxed = new Box(0);

for (let i = 0; i <= 999999999; i++) {
  oldboxed.write(boxed.read());
  boxed.write(i);
}

console.log(oldboxed.read());
console.log(boxed.read());
```

**Comment by vrescobar** [comment]
I don't want to discourage but I put together the sources for the new 2.4.0 shen kernel[1] with the latest changes announced by Mark in head hoping that with the new release everything would be smoother and the times triplicated (in fact the chrome tab became irresponsive the first time I loaded it). Desktop is at 24 seconds and mobile at 39 seconds.

If I activate the production version at webpack, curiously enough Chrome desktop loads in 1.8 seconds, safari crashes and firefox desktop takes 33 seconds; mobile did not even load (I guess the OS killed the page).


[1] The sources are in the repo, but seems that there is some broken link: https://github.com/Shen-Language/shen-sources/issues/76

**Comment by rkoeninger** [comment]
2.4.0 is the version of shen-cl. Last Shen kernel release was 21.1

You're saying you built klambda for what's on shen-sources/master and then built ShenScript with that and it's slower?

Here's the diff, I don't know why it would be so much slower: https://github.com/Shen-Language/shen-sources/compare/4b17acab66e7c4d7d0d26f8b82fb11bee36a52a9...master

**Comment by vrescobar** [comment]
I made more or less something like that (I wanted ShenScript to pick always the latest)

```
git clone https://github.com/Shen-Language/shen-sources kernel
pushd kernel && make pure $$ make klambda Shen=/usr/local/bin/shen-sbcl && make release && popd
npm run lint || true
npm run test-backend
npm run render-kernel
npm run test-kernel
npm run test-frontend
```

**Comment by vrescobar** [comment]
I retried with the newest shen-sbcl 2.5.0 (just in case the generated klambda sources could make any difference) and:


```
-------------------------------------------
| sync  | 1 failed     | 14s, 539ms       |
| async | all passed   | 45s, 53ms        |
| both  |              | 59s, 592ms       |
-------------------------------------------
```

Time chrome: environment created in 1286ms.
Time firefox: `A page is slowing down your browser. What would you like to do?` then 27465ms.
Time safari: environment created in 24004ms.
Time safari (iOS): (still waiting)

The problem is that running the page with a performance profiler makes it even slower and I think firefox just won't end (gave up after 9 minutes waiting for it). But even if I run it on chrome I can get some valuable information:

<img width="419" alt="Screenshot 2019-09-01 at 17 09 19" src="https://user-images.githubusercontent.com/6759126/64078457-4dff6c80-ccdb-11e9-872b-d65888d63ea7.png">

Chrome performance profiler points out that most of the CPU time, about 40%, is spent  **in the internal javascript eval function** which most probably means that the cause of the slow down must have been some of the commits from the previous days (fast on chrome, slow everywhere else).

Which brought me to retry the test with an older commit (c47328a95b8011c9bdbe1979561aa90247c1b7bd) but still the newest shen sources. That leads to similar test times:

```
--------------------------------
| sync  | 13s, 205ms           |
| async | 43s, 354ms           |
| both  | 56s, 559ms           |
--------------------------------
```
Using webpack in development mode:

Chrome: environment created in 1385ms.
Firefox: environment created in 13095ms.
Safari: environment created in 23509ms.
Safari (iOS): (never shows the message)

Conclusion:
Test times are not reliable measure for boot times.  Somewhere between d849ea688545f4c12ce1489afbd732945efa3ca0 and c47328a95b8011c9bdbe1979561aa90247c1b7bd the performance got a serious kick and by reading the conversation and the performance profiler I guess it **might** be the usage of local variables but I cannot tell for sure.

Browser tests: might make sense to make slower tests with an array of [selenium drivers](https://github.com/SeleniumHQ/selenium/tree/master/javascript/node/selenium-webdriver) but I never really tried in an actual project.

**Comment by rkoeninger** [comment]
For those two commits, the only main difference is the `trap` functions were replaced with try/catch in iifes: https://github.com/rkoeninger/ShenScript/compare/c47328a95b8011c9bdbe1979561aa90247c1b7bd...d849ea688545f4c12ce1489afbd732945efa3ca0

**Comment by rkoeninger** [comment]
Hoisting idle symbol and global references helped performance a good amount:

| scenario | before | after |
|:-|:-----|:-----|
| test suite (sync) | 11s | 7s |
| test suite (async) | 1m 4s | 56s |
| async web startup (brave) | 1200ms | 950ms |
| async web startup (firefox) | 22s | 12s |
| sync web startup (brave) | 800ms | 550ms |
| sync web startup (firefox) | 1800ms | 1200ms |

Firefox still slower than Brave (Chromium), and sync is still much faster than async, but measurable benefits all around.

**Comment by rkoeninger** [comment]
Idle symbol hoisting still not done for symbols starting with non-alpha character, might be certain symbol(s) that gets broken by that process and needs to be excluded explicitly?

**Comment by rkoeninger** [comment]
Adding overrides has rendered another significant improvement:

| scenario | before | after |
|:-|:-----|:-----|
| test suite (sync) | 7s | 4s |
| test suite (async) | 56s | 39s |
| async web startup (brave) | 950ms | 400ms |
| async web startup (firefox) | 15s | 3s |
| sync web startup (brave) | 550ms | 280ms |
| sync web startup (firefox) | 1200ms | 500ms |

**Comment by tizoc** [comment]
@rkoeninger looking great. Pre-compiling `declare` and other things that use `eval-kl` will probably improve the startup time a good amount too.

**Comment by tizoc** [comment]
Just checked the overrides, most of those are things that make the Shen compiler faster (dicts, `variable?`, etc), which kind of confirms my suspicion that at startup a lot of the time is being spent calling Shen's `eval`.

**Comment by tizoc** [comment]
The precompilation I'm talking about is https://github.com/Shen-Language/shen-sources/issues/83 btw, I haven't done anything on that front yet, but it is planned.

**Comment by rkoeninger** [comment]
Using otabat's improved types.kl takes that startup time much lower (little effect on test suite time as expected):

| scenario | before | after |
|:-|:-----|:-----|
| sync web startup (brave) | 280ms | 130ms |
| async web startup (brave) | 400ms | 60ms |
| sync web startup (firefox) | 500ms | 270ms |
| async web startup (firefox) | 3s | 170ms |

Although, the fact that the async time is now shorter than the sync time is suspicious. I'm guessing there are promises that aren't being appropriately awaited, allowing the startup to return before it's actually finished, but I'm not sure where.

**Comment by tizoc** [comment]
Yes, the async times are suspicious, but the sync improvements confirms what I was saying.

**Comment by rkoeninger** [comment]
Upgrading to kernel 22.0 has yielded the performance benefits seen when testing with the hand-optimized types.kl.

I'm questioning the need to have a sync mode now that async startup performance is so good, even in Firefox.

**Comment by tizoc** [comment]
@rkoeninger so, are the times from your previous post correct? async is faster than sync?

**Comment by rkoeninger** [comment]
I'm looking at that now. Especially in Firefox, you can see that it prints "done" after about 150ms and but the activity indicator doesn't stop for another ~500ms. There's probably an `await` missing somewhere that I just never noticed.

**Comment by rkoeninger** [comment]
I think the time difference is the time it takes for the definitions in `frontend` to complete. They're not being awaited.

**Comment by rkoeninger** [comment]
Alright, with all the declarations in `frontend` being properly awaited, I get async startup time of 200ms (Brave) and 1500ms (Firefox).

I think the proscribed way of running Shen in the browser will be to have it run in the background in a worker, so the page won't stall for 1.5s or anything, but 1.5s is still not great.

The `frontend` declarations are doing a long series of `declare`s, calling `declare` from JS, that could be optimized in a similar way to the `expand-dynamic` extension.

**Comment by rkoeninger** [comment]
Optimized the `defineTyped` function in `frontend.js` the same way `expand-dynamic` does and got back to the 60ms (Chrome), 140ms (Firefox) startup times.

**Comment by rkoeninger** [comment]
@vrescobar Can try do the iOS measurement again so we get a side-by-side comparison?

**Comment by tizoc** [comment]
@rkoeninger how much of the startup time is spent on `shen.initialise` + starting REPL related stuff? (just trying to get an idea of how much overhead is in there for programs that don't use interactive features)

**Comment by rkoeninger** [comment]
I don't have measurements for that, but prior experience makes me think the time is almost all in `shen.initialise` and since the startup time refers to web startup time, there's no repl being started. Running the `defun`s has always taken a trivial amount of time.

**Comment by tizoc** [comment]
Ok, good to know (that was expected, but just wanted to be sure).

There are parts of initialise that you can skip if you are not launching the REPL.

`shen.initialise` calls these four functions, and only the first two are required, the last two are for typechecking:

- `shen.initialise-environment`sets most globals, most of which are not needed unless you are using Shen's `eval`, but there are some things here that are needed for running programs, maybe I could look into splitting this one for the next release.
- `shen.initialise-lambda-forms` associates lambda forms to function names
- `shen.initialise-signedfunc-lambda-forms` same but for signature functions used for typechecking
- `shen.initialise-signedfuncs` stores the signatures of functions

**Comment by tizoc** [comment]
I'm also thinking that both `initialise-*-lambda-forms` functions can probably be re-defined statically in Javascript too which may save some extra time. At the least, with the "lambdas" being hoisted so that they are not created inside that function (the function would just associate the names).

For example:

`(shen.set-lambda-form-entry (cons shen.datatype-error (lambda X (shen.datatype-error X))))`

becomes:

`(shen.set-lambda-form-entry (cons shen.datatype-error $referenceToHoistedLambda))`.

I don't know if this makes a difference for the Javascript VMs, maybe when the compiler sees that there are no external references in the lambdas it takes care of this itself and avoids the allocation, but I'm not sure.

I should probable update this in the kernel to avoid the `cons`es, it is how it was originally, but it is not needed now when using the `expand-dynamic` extension. Not sure how much that will help, but it will not hurt.

**Comment by rkoeninger** [comment]
I experimented un-wrapping "wrapper lambdas", ran into some issue, it wasn't working, started to question how much it would help, set it aside.

It could also introduce weird arity issues where a function is arity 2 but it's wrapped inside 2 layers of lambda, each arity 1.

As far as the note on the one commit about package size, I plan to remove the sync rendering of the kernel and sync mode, which will cut the deployable size by almost half.

**Comment by tizoc** [comment]
Note that what I mentioned in the previous comment is not about unwrapping anything, it is about "creating" the lambdas at the top level, and not during initialise. It may or may not make a difference to the JS engine, depending on if it is making closures or not when the functions are created inside another function.

An easy way to know is to skip the `*-lambda-forms` functions and see how much that speedups startup, if it makes no difference then thats not where time is being spent.

**Comment by tizoc** [comment]
Regarding wrapper lambdas, I tried that in Shen/Scheme (for cases where the arity = 1), turns out that it doesn't work because sometimes lambdas are created before the function is defined, so you cannot unwrap them without getting an invalid reference. It happens a lot inside the kernel because lambda forms for macros are registered before the macro function is evaluated, and other similar situations. Looking into that is something I have left for later.

**Comment by rkoeninger** [comment]
> it doesn't work because sometimes lambdas are created before the function is defined

Right, I ran into that. And there's no point trying to get around it because if you wrap the function reference in a lookup function, then you're re-creating the same problem. Except maybe an opportunity to avoid async and the async test suite run does take 10x longer.

I don't know if hoisting lambdas would help a whole lot - it shouldn't take a JS engine any longer to create a lambda instance than any other value.

I did just push a change that flattens rendered lambdas. Very small performance improvement (might be within variation), but produces slightly more compact code and I like it.

**Comment by rkoeninger** [comment]
Providing overrides for `put`, `shen.initialise_arity_table` and `shen.active-cons` have reduced web startup time from a minimum observed of 50ms to 40ms. The `shen.initialise_arity_table` assumes there are no entries in the `*property-vector*` yet for those functions. Also, if `shen.initialise_arity_table` is overridden, `put` might not need to be overridden.

I overrode `shen.active-cons`, but now `shen.curry-type-h` is taking up time, I might do `shen.curry-type` as a single override. The fact that these functions are non-tail recursive means there will be a lot of trampolining interlaced with promise chaining (bit waste of time).

Another profiling observation is that a decent amount of time is spent throwing and catching exceptions and the kernel would be faster if it didn't depend on exceptions for flow control.

**Comment by tizoc** [comment]
> Another profiling observation is that a decent amount of time is spent throwing and catching exceptions and the kernel would be faster if it didn't depend on exceptions for flow control.

See [this](https://github.com/tizoc/shen-scheme/blob/master/src/compiler.shen#L169-L188), may help.

**Comment by vrescobar** [comment]
With the latest master:

Desktop firefox: 30193ms
Desktop chrome: 1410ms.
Desktop safari: 23798ms.

mobile safari: 43992ms

**Comment by vrescobar** [comment]
New measurements with the HEAD of this repo, hash: 73d9461b626eaa8fefae844194d81718181e0a83 (actual latest master HEAD)

Desktop firefox: 166ms.
Desktop chrome: 104ms.
Desktop safari: 140ms.

mobile safari: 196ms.

The last weeks I've been busy with something personal and I did not follow how things developed (I think I read about an earlier expansion of klambda), and I am very impressed with the results!

The size grew to 1.2MB (184Kb when gzip), I wonder if we could go smaller than that while keeping the klambda expanded, bit now it is already usable for a realistic webpage (especially since I expect libraries to be very small and not grow much the total size).

In the other hand every single browser did not properly load the webpage on the first time I wrote the URL and had to reload it. My settings were that it should always remove the cache, so I cannot tell if it also kept an old cash before I reloaded it again; now it seems to work without any issues!

**Comment by tizoc** [comment]
> Desktop firefox: 166ms.
> Desktop chrome: 104ms.
> Desktop safari: 140ms.
> 
> mobile safari: 196ms.

Looking good!

> The size grew to 1.2MB (184Kb when gzip), I wonder if we could go smaller than that while keeping the klambda expanded, bit now it is already usable for a realistic webpage (especially since I expect libraries to be very small and not grow much the total size).

It may be possible but I don't think it is worth it. As I mentioned before, consider that the size you see there is for loading the whole compiler and interactive environment, and unless your program uses the REPL, eval, and the typechecker at runtime, most of that code doesn't have to be loaded (same applies for the platform's kl-><lang> compiler itself), Shen's runtime for static programs is quite small.

**Comment by rkoeninger** [comment]
@vrescobar Thanks for taking the time to get those measurements. Good to see the times are decent even on mobile.

As for the bundle size, I'm looking at removing the sync version of the kernel which would cut the size in half, and make the codebase simpler. Now that async mode is so fast, sync mode is not as necessary (but it is still 10x faster when running the test suite and 2x faster on web startup which is hard to throw away).

**Comment by tizoc** [comment]
~185ms on safari. Works, but for some reason setting console.log as the resolve function of the promise doesn't work, I have to use `alert` to see the result.

**Comment by rkoeninger** [comment]
Added babel plugin to webpack config to replace use of spread operator with Edge-compatible code. Now works in Edge as well. Load time ~250ms (compare 50ms in Chrome, 100ms in Firefox).

Speed isn't great but it works.

**Comment by rkoeninger** [comment]
Added functionality to the compiler to detect when a defun refers only to primitive (or overridden with strictly sync version) functions or itself and then render that function in sync mode. At least a 20% speedup.

At this point I deemed the async mode fast enough and dumped the pre-rendered sync kernel. Cut the web bundle and npm packages' size by about 45%.

Version 0.9.0 is the first to be [https://www.npmjs.com/package/shen-script](deployed to npm).

Time for this thread to close. Thanks, guys!
